File: packages/core/moxfield.js
'use strict';

function buildCollectionUrlFromId(collectionId) {
  return `https://moxfield.com/collection/${collectionId}`;
}

function isValidCollectionUrl(urlString) {
  try {
    const u = new URL(urlString);
    const isHttps = u.protocol === 'https:';
    const isTargetHost = u.hostname === 'moxfield.com' || u.hostname === 'www.moxfield.com';
    const isCollectionPath = /^\/collection\/[A-Za-z0-9_-]+\/?$/.test(u.pathname);
    return isHttps && isTargetHost && isCollectionPath;
  } catch {
    return false;
  }
}

function extractCollectionIdFromUrl(urlString) {
  try {
    const u = new URL(urlString);
    const match = u.pathname.match(/^\/collection\/([A-Za-z0-9_-]+)\/?$/);
    return match ? match[1] : null;
  } catch {
    return null;
  }
}

module.exports = {
  buildCollectionUrlFromId,
  isValidCollectionUrl,
  extractCollectionIdFromUrl
};

File: internal/moxfield/moxfield.go
package moxfield

import "net/url"

// BuildCollectionURLFromID constructs a canonical collection URL from an ID.
func BuildCollectionURLFromID(collectionID string) string {
	return "https://moxfield.com/collection/" + collectionID
}

// IsValidCollectionURL reports whether the input is a supported Moxfield collection URL.
func IsValidCollectionURL(urlString string) bool {
	u, err := url.Parse(urlString)
	if err != nil {
		return false
	}
	if u.Scheme != "https" {
		return false
	}
	if u.Hostname() != "moxfield.com" && u.Hostname() != "www.moxfield.com" {
		return false
	}
	return collectionPathRegex.MatchString(u.EscapedPath())
}

// IsValidCollectionID reports whether id is a well-formed collection ID.
// Valid IDs are 1–32 characters composed of ASCII letters, digits, underscores, or hyphens.
func IsValidCollectionID(id string) bool {
	return collectionIDRegex.MatchString(id)
}

// ExtractCollectionIDFromURL returns the collection ID from a valid collection URL.
func ExtractCollectionIDFromURL(urlString string) string {
	u, err := url.Parse(urlString)
	if err != nil {
		return ""
	}
	parts := collectionPathRegex.FindStringSubmatch(u.EscapedPath())
	if len(parts) != 2 {
		return ""
	}
	return parts[1]
}
